### PR TITLE
Hare certificates

### PR DESCRIPTION
**Pull request by gavraz**
* No certificate in a proposal
* Track certificates
* Validate status message only at the end of the round
* fixed relevant unit tests and integration tests accordingly

### CODE REVIEW

**Comment by gavraz** [comment]
This PR also fixes issues:
#404 
#425

File: hare/algorithm.go [hunk 1]
```diff
@@ -161,18 +164,21 @@ func roleFromRoundCounter(k uint32) Role {
...
 	}
 
+	// TODO: validate role proof
+
 	// validate role
-	if proc.oracle.Role(m.Message.K, Signature(m.Message.RoleProof)) != roleFromRoundCounter(m.Message.K) {
+	msgK := m.Message.K
+	proof := Signature(m.Message.RoleProof)
+	role := roleFromRoundCounter(m.Message.K)
+	if proc.oracle.Role(m.Message.K, proof) != role {
```
Comment (zalmen) [inline hare/algorithm.go]: use msgK instead of m.Message.K

File: hare/algorithm.go [hunk 2]
```diff
@@ -161,18 +164,21 @@ func roleFromRoundCounter(k uint32) Role {
...
 		log.Warning("Invalid role detected. Expected: %v Actual: %v",
-			proc.oracle.Role(proc.k, Signature(m.Message.RoleProof)), roleFromRoundCounter(m.Message.K))
+			proc.oracle.Role(msgK, proof), role)
```
Comment (zalmen) [inline hare/algorithm.go]: replace proc.oracle.Role(msgK, proof) with a var (it is calculated twice)

File: hare/algorithm.go [hunk 3]
```diff
@@ -344,21 +350,26 @@ func (proc *ConsensusProcess) processCommitMsg(msg *pb.HareMessage) {
 }
 
 func (proc *ConsensusProcess) processNotifyMsg(msg *pb.HareMessage) {
+	s := NewSet(msg.Message.Values)
+	proc.notifyTracker.OnCertificate(msg.Cert.AggMsgs.Messages[0].Message.K, s)
```
Comment (zalmen) [inline hare/algorithm.go]: move it to inside OnNotify

File: hare/messagevalidation.go [hunk 4]
```diff
@@ -46,18 +47,39 @@ func (validator *MessageValidator) ValidateMessage(m *pb.HareMessage, k uint32)
...
-	isSameIteration := iterationFromCounter(k) == iterationFromCounter(m.Message.K)
-	currentRound := k % 4
 	switch MessageType(m.Message.Type) {
 	case PreRound:
 		return true
+	case Notify:
+		return true
+	}
+
+	expIter := iterationFromCounter(k)
```
Comment (zalmen) [inline hare/messagevalidation.go]: replace all of this code (till the end of the switch) with a check whether k == m.Message.K

File: hare/statustracker.go [hunk 5]
```diff
@@ -41,17 +43,35 @@ func (st *StatusTracker) RecordStatus(msg *pb.HareMessage) {
...
-		st.maxKi = msg.Message.Ki
-		st.maxRawSet = msg.Message.Values
+	st.statuses[pub.String()] = msg
+}
+
+func (st *StatusTracker) AnalyzeStatuses(isValid func(m *pb.HareMessage) bool) {
+	for key, m := range st.statuses {
+		if !isValid(m) { // only keep valid messages
+			delete(st.statuses, key)
+			continue
```
Comment (zalmen) [inline hare/statustracker.go]: merge the two for loops

**Review by zalmen — CHANGES_REQUESTED** [review]

File: hare/algorithm.go (hunk 1 above)
Comment (gavraz) [reply hare/algorithm.go]: Fixed.

File: hare/algorithm.go (hunk 2 above)
Comment (gavraz) [reply hare/algorithm.go]: Fixed

File: hare/algorithm.go (hunk 3 above)
Comment (gavraz) [reply hare/algorithm.go]: Fixed

File: hare/messagevalidation.go (hunk 4 above)
Comment (gavraz) [reply hare/messagevalidation.go]: Fixed

File: hare/statustracker.go (hunk 5 above)
Comment (gavraz) [reply hare/statustracker.go]: Fixed

**Review by zalmen — APPROVED** [review]

**Pull request merged**